File: app/src/main/java/com/drdisagree/iconify/Iconify.java
package com.drdisagree.iconify;

import android.app.Application;
import android.content.Context;

import com.drdisagree.iconify.utils.helper.LocaleHelper;
import com.google.android.material.color.DynamicColors;

import java.lang.ref.WeakReference;

public class Iconify extends Application {

    private static Iconify instance;
    private static WeakReference<Context> contextReference;

    public void onCreate() {
        super.onCreate();
        instance = this;
        contextReference = new WeakReference<>(getApplicationContext());
        DynamicColors.applyToActivitiesIfAvailable(this);
    }

    public static Context getAppContext() {
        if (contextReference == null || contextReference.get() == null) {
            contextReference = new WeakReference<>(Iconify.getInstance().getApplicationContext());
        }

        return contextReference.get();
    }

    public static Context getAppContextLocale() {
        return LocaleHelper.setLocale(getAppContext());
    }

    private static Iconify getInstance() {
        return instance;
    }
}

File: app/src/main/java/com/drdisagree/iconify/app/Iconify.kt
package com.drdisagree.iconify.app

import android.app.Application
import android.content.Context
import dagger.hilt.android.HiltAndroidApp
import java.lang.ref.WeakReference

@HiltAndroidApp
class Iconify : Application() {

    override fun onCreate() {
        super.onCreate()

        _instance = this
        _contextReference = WeakReference(applicationContext)
    }

    companion object {

        private lateinit var _instance: Iconify
        private lateinit var _contextReference: WeakReference<Context>

        private var instance: Iconify
            get() {
                if (this::_instance.isInitialized.not()) {
                    _instance = Iconify()
                }
                return _instance
            }
            set(value) {
                _instance = value
            }

        val appContext: Context
            get() {
                if (this::_contextReference.isInitialized.not() || _contextReference.get() == null) {
                    _contextReference = WeakReference(instance.applicationContext)
                }
                return _contextReference.get()!!
            }
    }
}